Generify Palettes Implementation

gabizou opened the pull request:
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/2244) | **SpongeCommon**

This will bring the implementation of Palettes to be more inline with their mutability/immutability aspects, though I still have some questions to answer:
- [ ] How do we want to handle a `Palette` being built by referring to a Minecraft owned Palette (like the case of BlockStates)
- [ ] How do we want to make a `Palette` based on a Minecraft registry?
- [ ] Should we always make immutable palettes when presented with a registry for the string serialization handling?

More to come as it gets fleshed out.

Cybermaxke commented on src/main/java/org/spongepowered/common/world/schematic/ImmutableBimapPalette.java:
```diff
@@ -0,0 +1,107 @@
+    private final int maxId;
+
+    public ImmutableBimapPalette(final PaletteType<T> paletteType, final BiMap<Integer, T> reference) {
+        final ImmutableBiMap.Builder<Integer, T> builder = ImmutableBiMap.builder();
```
Why not use an `Int2ObjectMap` and `Object2IntMap` if you're going to make a copy?

kashike replied:
:+1: for using fastutil here

Zidane replied:
Agreed

gabizou committed:
Introduce MemoizedSupplier for Supplier memoization

A Supplier normally would evaluate every single time being called, but in
some cases, we'd rather be able to memoize the value generated from the
Supplier after the first call. This is not meant to be leaked beyond memoizing
global objects that are referentially transparent themselves (like an object
being comprised of functions/suppliers that rely on other global variables).

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Reintroduce Blockstate(De)Serializer

Traditionally, Mojang is appearing to want to store a BlockState as a map to better
suit DFU's ability to update a state's values and properties. This map based
representation however, is not ideal to being used for storage with a palette that
retains the integer to string like mapping internally. It may well be possible to
take advantage of DFU in the future, but at this time, a less breakage of the
Schematic Specification is required.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Implement new Palettes

Palettes are much more streamlined now, with global palettes being memoized functions
built off the global registries (meaning they are reload-aware) while copies can be
isolated by themselves and rebuilt however need be.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>
